### PR TITLE
[InstCombine] Add folds for (add/sub/disjoint_or/icmp C, (ctpop (not x)))

### PR DESCRIPTION
**Pull request by goldsteinn**
- [InstCombine] Add tests for folding `(add/sub/disjoint_or/icmp C, (ctpop (not x)))`; NFC
- [InstCombine] Add folds for `(add/sub/disjoint_or/icmp C, (ctpop (not x)))`
----
`(ctpop (not x))` <-> `(sub nuw nsw BitWidth(x), (ctpop x))`. The
`sub` expression can sometimes be constant folded depending on the use
case of `(ctpop (not x))`.

This patch adds fold for the following cases:

`(add/sub/disjoint_or C, (ctpop (not x))`
    -> `(add/sub/disjoint_or C', (ctpop x))`
`(cmp pred C, (ctpop (not x))`
    -> `(cmp swapped_pred C', (ctpop x))`

Where `C'` depends on how we constant fold `C` with `BitWidth(x)` for
the given opcode.

Proofs: https://alive2.llvm.org/ce/z/qUgfF3

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:776]:
```suggestion
  Value *Op;
  if (!match(I->getOperand(1 - ConstIdx), m_OneUse(m_Intrinsic<Intrinsic::ctpop>(m_Value(X)))))
    return nullptr;
```

Missing one-use constraints for ctpop?

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp]: Fixed + added test.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:814]: Looks like it doesn't hold for `ctpop.i2`.
Alive2: https://alive2.llvm.org/ce/z/CiYgW4

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:814]: Ah, good catch. The lemma also requires C is positive.

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:814]: i just disabled to signed predicates. Shouldn't make a difference as any comparison of `ctpop, C` should be trivial to simplify in unsigned form.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for approval from @nikic.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:791]: I think something is missing after "if this check is"?

**Review by nikic — APPROVED** [review]
LGTM

**Pull request closed without merging**
